File: agents/go/resource_test.go
package openlog

import (
	"context"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"testing"
)

// fixture writes files below a temp root, like the infra agent's hostfstest.
func fixture(t *testing.T, files map[string]string) string {
	t.Helper()
	root := t.TempDir()
	for p, content := range files {
		full := filepath.Join(root, p)
		if err := os.MkdirAll(filepath.Dir(full), 0o755); err != nil {
			t.Fatal(err)
		}
		if err := os.WriteFile(full, []byte(content), 0o644); err != nil {
			t.Fatal(err)
		}
	}
	return root
}

func noPlatformID(t *testing.T) {
	old := platformHostID
	platformHostID = func(context.Context) string { return "" }
	t.Cleanup(func() { platformHostID = old })
}

// The fixtures mirror agents/infra/internal/resource TestHostIDChain so both agents resolve
// the same host.id on the same machine.
func TestHostIDChainMatchesInfraAgent(t *testing.T) {
	noPlatformID(t)
	ctx := context.Background()
	infra := "/var/lib/openlog-infra-agent"
	cases := []struct {
		name, want, source string
		files              map[string]string
	}{
		{"machine-id", "0123456789abcdef0123456789abcdef", "/etc/machine-id", map[string]string{
			"/etc/machine-id": "0123456789abcdef0123456789abcdef\n", "/var/lib/dbus/machine-id": "ffffffffffffffffffffffffffffffff\n"}},
		{"dbus fallback lower-cased", "abcdef0123456789abcdef0123456789", "/var/lib/dbus/machine-id", map[string]string{
			"/etc/machine-id": "\n", "/var/lib/dbus/machine-id": "ABCDEF0123456789abcdef0123456789\n",
			"/sys/class/dmi/id/product_uuid": "11111111-2222-3333-4444-555555555555\n"}},
		{"dmi fallback", "11111111-2222-3333-4444-555555555555", "/sys/class/dmi/id/product_uuid", map[string]string{
			"/sys/class/dmi/id/product_uuid": "11111111-2222-3333-4444-555555555555\n"}},
		{"too short machine-id ignored", "11111111-2222-3333-4444-555555555555", "/sys/class/dmi/id/product_uuid", map[string]string{
			"/etc/machine-id": "abc\n", "/sys/class/dmi/id/product_uuid": "11111111-2222-3333-4444-555555555555\n"}},
		{"zero dmi ignored, infra state file used", "9b2c7d7e-1f0a-4c1e-9a55-6a3f1c2b8d10", hostIDSourceInfra, map[string]string{
			"/sys/class/dmi/id/product_uuid":              "00000000-0000-0000-0000-000000000000\n",
			"/var/lib/openlog-infra-agent/host-id":        "9b2c7d7e-1f0a-4c1e-9a55-6a3f1c2b8d10\n",
			"/var/lib/openlog-infra-agent/host-id.tmp":    "garbage",
			"/var/lib/openlog-infra-agent/unrelated-file": "x"}},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			root := fixture(t, tc.files)
			id, src := resolveHostID(ctx, hostFS{root: root}, "", infra, t.TempDir())
			eq(t, id, tc.want)
			eq(t, src, tc.source)
		})
	}

	t.Run("generated and persisted", func(t *testing.T) {
		root := fixture(t, map[string]string{"/etc/machine-id": "\n"})
		state := t.TempDir()
		id, src := resolveHostID(ctx, hostFS{root: root}, "", infra, state)
		if !regexp.MustCompile(`^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`).MatchString(id) {
			t.Fatalf("generated id %q", id)
		}
		eq(t, src, hostIDSourceGenerated)
		again, _ := resolveHostID(ctx, hostFS{root: root}, "", infra, state)
		eq(t, again, id)
	})
}

func TestContainerID(t *testing.T) {
	const id = "3f4b1a2c5d6e7f8091a2b3c4d5e6f708192a3b4c5d6e7f8091a2b3c4d5e6f7a8"
	cases := []struct {
		name  string
		files map[string]string
		want  string
	}{
		{"cgroup v1 docker", map[string]string{"/proc/self/cgroup": "12:memory:/docker/" + id + "\n11:cpu:/docker/" + id + "\n"}, id},
		{"cgroup v2 systemd scope", map[string]string{"/proc/self/cgroup": "0::/system.slice/docker-" + id + ".scope\n"}, id},
		{"kubernetes containerd", map[string]string{"/proc/self/cgroup": "0::/kubepods.slice/kubepods-burstable.slice/kubepods-pod1.slice/cri-containerd-" + id + ".scope\n"}, id},
		{"cgroup namespace uses mountinfo", map[string]string{
			"/proc/self/cgroup": "0::/\n",
			"/proc/self/mountinfo": "600 580 0:50 / / rw,relatime - overlay overlay rw\n" +
				"612 600 254:1 /docker/containers/" + id + "/resolv.conf /etc/resolv.conf rw,relatime - ext4 /dev/vda1 rw\n"}, id},
		{"sandbox mount skipped", map[string]string{
			"/proc/self/cgroup":    "0::/\n",
			"/proc/self/mountinfo": "1 0 0:1 /var/lib/containerd/io.containerd.grpc.v1.cri/sandboxes/" + id + "/hostname /etc/hostname rw - ext4 /dev/sda rw\n"}, ""},
		{"not in a container", map[string]string{"/proc/self/cgroup": "0::/user.slice/user-1000.slice/session-2.scope\n"}, ""},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			eq(t, containerID(hostFS{root: fixture(t, tc.files)}), tc.want)
		})
	}
}

func TestK8sAttributes(t *testing.T) {
	root := fixture(t, map[string]string{"/var/run/secrets/kubernetes.io/serviceaccount/namespace": "shop\n"})
	got := k8sAttributes(hostFS{root: root}, env(map[string]string{
		"KUBERNETES_SERVICE_HOST": "10.0.0.1", "HOSTNAME": "api-7d9f-abc", "K8S_NODE_NAME": "node-1", "K8S_POD_UID": "uid-1",
	}))
	eq(t, got["k8s.pod.name"], "api-7d9f-abc")
	eq(t, got["k8s.namespace.name"], "shop")
	eq(t, got["k8s.node.name"], "node-1")
	eq(t, got["k8s.pod.uid"], "uid-1")
	if len(k8sAttributes(hostFS{root: root}, env(map[string]string{"HOSTNAME": "laptop"}))) != 0 {
		t.Error("k8s attributes outside a pod")
	}
}

func TestBuildResource(t *testing.T) {
	noPlatformID(t)
	root := fixture(t, map[string]string{
		"/etc/machine-id":            "0123456789abcdef0123456789abcdef\n",
		"/etc/os-release":            "ID=debian\nVERSION_ID=\"12\"\nPRETTY_NAME=\"Debian GNU/Linux 12 (bookworm)\"\n",
		"/proc/sys/kernel/hostname":  "web-01\n",
		"/proc/sys/kernel/osrelease": "6.1.0-18-amd64\n",
		"/proc/sys/kernel/arch":      "x86_64\n",
	})
	cfg, _, err := loadConfig(env(map[string]string{
		"OPENLOG_HOST_ROOT": root, "OPENLOG_SERVICE_NAME": "checkout", "OPENLOG_SERVICE_VERSION": "1.2.3",
		"OPENLOG_ENVIRONMENT": "production", "OPENLOG_SERVICE_NAMESPACE": "shop", "OPENLOG_STATE_DIR": t.TempDir(),
		"OPENLOG_RESOURCE_ATTRIBUTES": "team=payments,service.name=ignored,host.name=override-name",
	}), nil)
	if err != nil {
		t.Fatal(err)
	}
	res, source, err := buildResource(context.Background(), cfg, env(nil))
	if err != nil {
		t.Fatal(err)
	}
	got := map[string]string{}
	for _, kv := range res.Attributes() {
		got[string(kv.Key)] = kv.Value.Emit()
	}
	want := map[string]string{
		"service.name": "checkout", "service.version": "1.2.3", "service.namespace": "shop",
		"deployment.environment.name": "production", "telemetry.distro.name": "openlog", "telemetry.distro.version": Version,
		"telemetry.sdk.language": "go", "team": "payments", "host.name": "override-name", "process.runtime.name": "go",
	}
	if runtime.GOOS == "linux" {
		for k, v := range map[string]string{"host.id": "0123456789abcdef0123456789abcdef", "os.name": "debian", "os.version": "12",
			"os.description": "Debian GNU/Linux 12 (bookworm)", "host.arch": "amd64", "openlog.os.kernel_release": "6.1.0-18-amd64", "os.type": "linux"} {
			want[k] = v
		}
	}
	// host.id is resolved from the fixture on every OS (the file chain is not Linux-specific).
	want["host.id"] = "0123456789abcdef0123456789abcdef"
	eq(t, source, "/etc/machine-id")
	for k, v := range want {
		if got[k] != v {
			t.Errorf("%s = %q, want %q", k, got[k], v)
		}
	}
	if got["process.pid"] == "" || got["process.executable.name"] == "" {
		t.Errorf("process attributes missing: %v", got)
	}

	// An explicit host id wins over detection and user attributes.
	cfg.HostID = "explicit"
	cfg.ResourceAttributes["host.id"] = "from-attrs"
	res, source, _ = buildResource(context.Background(), cfg, env(nil))
	v, _ := res.Set().Value("host.id")
	eq(t, v.AsString(), "explicit")
	eq(t, source, hostIDSourceConfig)
}

File: agents/node/test/unit/resource.test.ts
import assert from 'node:assert/strict';
import * as fs from 'node:fs';
import * as os from 'node:os';
import * as path from 'node:path';
import { after, test } from 'node:test';
import { loadConfig } from '../../src/config';
import { buildResource, containerId, HostFS, k8sAttributes, normalizeArch, parseOSRelease, platform, resolveHostId } from '../../src/resource';

const tmpRoots: string[] = [];
function root(files: Record<string, string>): string {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'openlog-node-res-'));
  tmpRoots.push(dir);
  for (const [p, content] of Object.entries(files)) {
    fs.mkdirSync(path.join(dir, path.dirname(p)), { recursive: true });
    fs.writeFileSync(path.join(dir, p), content);
  }
  return dir;
}
after(() => {
  for (const d of tmpRoots) fs.rmSync(d, { recursive: true, force: true });
});

const CID = 'a'.repeat(12) + '0123456789abcdef'.repeat(3) + 'b'.repeat(4);

test('host id chain (same as the infra and Go agents)', () => {
  const origPlatform = platform.hostId;
  platform.hostId = () => 'platform-uuid-1234';
  try {
    // 0. infra agent runtime file wins
    let r = root({ '/run/openlog-infra-agent/host-id': 'infra-published-id\n', '/etc/machine-id': 'ABCDEF0123456789\n' });
    assert.deepEqual(resolveHostId(new HostFS(r), '/run/openlog-infra-agent', '/var/lib/openlog-infra-agent', path.join(r, 'state'), true), ['infra-published-id', 'infra-agent']);
    // 1. machine-id lower-cased; all-zero ids skipped
    r = root({ '/etc/machine-id': '00000000000000000000\n', '/var/lib/dbus/machine-id': 'ABCDEF0123456789\n' });
    assert.deepEqual(resolveHostId(new HostFS(r), '/run/openlog-infra-agent', '', path.join(r, 'state'), true), ['abcdef0123456789', '/var/lib/dbus/machine-id']);
    // invalid (short) ids skipped → product_uuid
    r = root({ '/etc/machine-id': 'short', '/sys/class/dmi/id/product_uuid': 'EC2A1B2C-0000-1111-2222-333344445555' });
    assert.deepEqual(resolveHostId(new HostFS(r), '', '', path.join(r, 'state'), true), ['ec2a1b2c-0000-1111-2222-333344445555', '/sys/class/dmi/id/product_uuid']);
    // 2. infra agent state dir
    r = root({ '/var/lib/openlog-infra-agent/host-id': 'generated-by-infra-1\n' });
    assert.deepEqual(resolveHostId(new HostFS(r), '/run/openlog-infra-agent', '/var/lib/openlog-infra-agent', path.join(r, 'state'), true), ['generated-by-infra-1', 'infra-agent-state']);
    // 3. platform id only off Linux
    r = root({});
    assert.deepEqual(resolveHostId(new HostFS(r), '', '', path.join(r, 'state'), false), ['platform-uuid-1234', 'platform']);
    // 4. generated and persisted
    const [gen, src] = resolveHostId(new HostFS(r), '', '', path.join(r, 'state'), true);
    assert.equal(src, 'generated');
    assert.match(gen, /^[0-9a-f-]{36}$/);
    assert.deepEqual(resolveHostId(new HostFS(r), '', '', path.join(r, 'state'), true), [gen, 'generated']);
  } finally {
    platform.hostId = origPlatform;
  }
});

test('container id from cgroup v1/v2 and mountinfo', () => {
  assert.equal(containerId(new HostFS(root({ '/proc/self/cgroup': `12:pids:/docker/${CID}\n0::/\n` }))), CID);
  assert.equal(containerId(new HostFS(root({ '/proc/self/cgroup': `0::/system.slice/docker-${CID}.scope\n` }))), CID);
  assert.equal(containerId(new HostFS(root({ '/proc/self/cgroup': `0::/kubepods/besteffort/pod1/cri-containerd-${CID}.scope` }))), CID);
  const mountinfo = [
    `100 90 0:50 /var/lib/containerd/io.containerd.grpc.v1.cri/sandboxes/${'c'.repeat(64)}/hostname /etc/hostname rw - ext4 /dev/vda1 rw`,
    `101 90 254:1 /docker/containers/${CID}/resolv.conf /etc/resolv.conf rw,relatime - ext4 /dev/vda1 rw`,
  ].join('\n');
  assert.equal(containerId(new HostFS(root({ '/proc/self/cgroup': '0::/\n', '/proc/self/mountinfo': mountinfo }))), CID);
  assert.equal(containerId(new HostFS(root({ '/proc/self/cgroup': '0::/user.slice\n' }))), '');
});

test('helpers', () => {
  assert.equal(normalizeArch('x64'), 'amd64');
  assert.equal(normalizeArch('aarch64'), 'arm64');
  assert.equal(normalizeArch('ia32'), 'x86');
  assert.deepEqual(parseOSRelease('# c\nID=debian\nVERSION_ID="12"\nPRETTY_NAME="Debian GNU/Linux 12 (bookworm)"\n'), {
    ID: 'debian',
    VERSION_ID: '12',
    PRETTY_NAME: 'Debian GNU/Linux 12 (bookworm)',
  });
  const r = root({ '/var/run/secrets/kubernetes.io/serviceaccount/namespace': 'shop\n' });
  assert.deepEqual(k8sAttributes(new HostFS(r), { KUBERNETES_SERVICE_HOST: '10.0.0.1', HOSTNAME: 'web-7d9', K8S_NODE_NAME: 'node-1' }), {
    'k8s.namespace.name': 'shop',
    'k8s.pod.name': 'web-7d9',
    'k8s.node.name': 'node-1',
  });
});

test('buildResource precedence and attributes', () => {
  const r = root({
    '/etc/machine-id': 'feedfacefeedface\n',
    '/proc/sys/kernel/hostname': 'apm-host\n',
    '/proc/sys/kernel/arch': 'aarch64\n',
    '/proc/sys/kernel/osrelease': '6.8.0-test\n',
    '/etc/os-release': 'ID=alpine\nVERSION_ID=3.22.0\nPRETTY_NAME="Alpine Linux v3.22"\n',
    '/proc/self/cgroup': `0::/docker/${CID}\n`,
  });
  const env = {
    OPENLOG_HOST_ROOT: r,
    OPENLOG_INFRA_RUNTIME_DIR: '/nonexistent-openlog',
    OPENLOG_SERVICE_NAME: 'checkout',
    OPENLOG_SERVICE_VERSION: '2.0.0',
    OPENLOG_ENVIRONMENT: 'prod',
    OTEL_RESOURCE_ATTRIBUTES: 'service.name=ignored,team=payments,host.name=override',
  };
  const { config } = loadConfig(env);
  const { resource, hostIdSource } = buildResource(config, env, { linux: true });
  const a = resource.attributes;
  assert.equal(hostIdSource, '/etc/machine-id');
  assert.equal(a['service.name'], 'checkout');
  assert.equal(a['service.version'], '2.0.0');
  assert.equal(a['deployment.environment.name'], 'prod');
  assert.equal(a['host.id'], 'feedfacefeedface');
  assert.equal(a['host.name'], 'override');
  assert.equal(a['host.arch'], 'arm64');
  assert.equal(a['os.type'], 'linux');
  assert.equal(a['os.name'], 'alpine');
  assert.equal(a['openlog.os.kernel_release'], '6.8.0-test');
  assert.equal(a['container.id'], CID);
  assert.equal(a['team'], 'payments');
  assert.equal(a['process.pid'], process.pid);
  assert.equal(a['process.runtime.name'], 'nodejs');
  assert.equal(a['telemetry.distro.name'], 'openlog');
  assert.equal(a['telemetry.sdk.language'], 'nodejs');
  assert.equal(a['telemetry.sdk.name'], 'opentelemetry');

  const explicit = loadConfig({ ...env, OPENLOG_HOST_ID: 'explicit-host-1' }).config;
  const res2 = buildResource(explicit, env, { linux: true });
  assert.equal(res2.resource.attributes['host.id'], 'explicit-host-1');
  assert.equal(res2.hostIdSource, 'config');
});
